### PR TITLE
OCPNODE-3372: Update defaultRuntime doc to show options and set default to crun

### PR DESCRIPTION
**Pull request by ngopalak-redhat**
This PR addresses an issue in the generated documentation for the defaultRuntime option

Previously, the documentation at https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/machine_apis/containerruntimeconfig-machineconfiguration-openshift-io-v1 did not list crun and runc as valid options for defaultRuntime.

While crun became the default runtime in OpenShift 4.18 (as per release notes: https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/release_notes/ocp-4-18-release-notes), the internal constant reflecting this default was not updated, though the functionality itself was correct.

This change updates the internal documentation comments and the relevant constant to accurately reflect:

Valid options: crun and runc
Default runtime: crun

Based on the review I have also added enum validation so that the failure can be caught earlier. Unit tests are added to ensure that it doesn't break any existing functionality.

### CODE REVIEW

File: machineconfiguration/v1/types.go [hunk 1]
```diff
@@ -844,7 +844,8 @@ type ContainerRuntimeConfiguration struct {
 	// +optional
 	OverlaySize *resource.Quantity `json:"overlaySize,omitempty"`
 
-	// defaultRuntime is the name of the OCI runtime to be used as the default.
+	// defaultRuntime is the name of the OCI runtime to be used as the default for containers.
+	// Options are runc and crun. Defaults to crun.
```
Comment (everettraven) [inline machineconfiguration/v1/types.go]: This field description should probably be updated to be a bit more consistent with the current way we document fields like this in OpenShift.

Something along the lines of:
```
defaultRuntime is an optional field used to configure the OCI runtime used by default for containers.
Allowed values are `runc`, `crun`, and omitted (`""`).
When set to `runc`, ...
When set to `crun`, ...
When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time. Currently, the default is `crun`.
```

File: machineconfiguration/v1/types.go [hunk 2]
```diff
@@ -844,7 +844,8 @@ type ContainerRuntimeConfiguration struct {
...
 	// +optional
 	DefaultRuntime ContainerRuntimeDefaultRuntime `json:"defaultRuntime,omitempty"`
```
Comment (everettraven) [inline machineconfiguration/v1/types.go:855]: I noticed that this field looks like it is enum restricted but doesn't have the markers to ensure that only valid values are provided at admission time. We should add the appropriate `kubebuilder:validation:Enum` marker either here at the field level or to the type alias to ensure we are validating the input early.

Technically the introduction of this would be a breaking change, so I'm curious - what is the behavior today if a user provides and invalid value to this field like `docker`?

File: machineconfiguration/v1/types.go [hunk 3]
```diff
@@ -855,7 +856,7 @@ const (
 	ContainerRuntimeDefaultRuntimeEmpty   = ""
 	ContainerRuntimeDefaultRuntimeRunc    = "runc"
 	ContainerRuntimeDefaultRuntimeCrun    = "crun"
-	ContainerRuntimeDefaultRuntimeDefault = ContainerRuntimeDefaultRuntimeRunc
+	ContainerRuntimeDefaultRuntimeDefault = ContainerRuntimeDefaultRuntimeCrun
```
Comment (yuqi-zhang) [inline machineconfiguration/v1/types.go:865]: Hmm, is this const actually being used somewhere? I don't think it's used in the MCO at all. Maybe we can drop it?

**Comment by ngopalak-redhat** [comment]
@everettraven and @yuqi-zhang . Thanks for your review. I'll make this a draft PR until I address those comments.

File: machineconfiguration/v1/types.go (hunk 3 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go:865]: Its used in MCO: https://github.com/openshift/machine-config-operator/blob/a4ec776a829dbb473a2091180a0e9537a6721a9e/pkg/controller/container-runtime-config/helpers.go#L621
Hence GoLand doesn't show the usage.

File: machineconfiguration/v1/types.go (hunk 2 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go:855]: This change introduces enum validation to the MCO operator's runtime configuration (now crun, runc, ""). Currently, specifying `docker` would lead to MCO operator failures. This is not a breaking change, as docker was not a supported value.
Reference: https://github.com/openshift/machine-config-operator/blob/a4ec776a829dbb473a2091180a0e9537a6721a9e/pkg/controller/container-runtime-config/helpers.go#L621

File: machineconfiguration/v1/types.go (hunk 1 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go]: Thanks, Updated it

File: machineconfiguration/v1/types.go [hunk 4]
```diff
@@ -844,18 +844,25 @@ type ContainerRuntimeConfiguration struct {
...
-	// defaultRuntime is the name of the OCI runtime to be used as the default.
+	// defaultRuntime is the name of the OCI runtime to be used as the default for containers.
+	// Allowed values are `runc`, `crun`, and omitted (`""`).
+	// When set to `runc`, OpenShift will use runc to execute the container
+	// When set to `crun`, OpenShift will use crun to execute the container
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default,
+	// which is subject to change over time. Currently, the default is `crun`.
+	// +kubebuilder:validation:Enum=crun;runc;""
 	// +optional
 	DefaultRuntime ContainerRuntimeDefaultRuntime `json:"defaultRuntime,omitempty"`
```
Comment (everettraven) [inline machineconfiguration/v1/types.go:855]: If we are allowing the empty string value, this will need to be a pointer, otherwise the `omitempty` will drop the empty string value on serialization.

Comment (everettraven) [reply machineconfiguration/v1/types.go:855]: Alternatively, if you don't need to distinguish between not provided and `""` (i.e not provided == `""`) you could drop the `omitempty` because you want the empty value to persist

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 5]
```diff
@@ -0,0 +1,80 @@
+apiVersion: apiextensions.k8s.io/v1 # Hack because controller-gen complains if we don't have this
+name: "ContainerRuntimeConfig"
+crdName: containerruntimeconfigs.machineconfiguration.openshift.io
+tests:
+  onCreate:
```
Comment (everettraven) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:5]: We should also add a test that setting it to an empty string and not providing the field at all acts as expected.

File: machineconfiguration/v1/types.go (hunk 4 above)
Comment (everettraven) [reply machineconfiguration/v1/types.go:855]: @JoelSpeed worked on a linter related to this recently so he may also have some more clear guidance here

Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go:855]: Thanks @everettraven . I have removed omitempty as "" is handled in the MCO code

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 5 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:5]: Thanks @everettraven . Added the test

**Comment by ngopalak-redhat** [comment]
> Because the logic previously existed in MCO that would cause failures if this isn't set correctly it should be fine to restrict this validation at admission time.
> 
> That being said, we should ensure that we properly test that this change ratchets correctly. Here is an example of a ratcheting validation test:
> 
> https://github.com/openshift/api/blob/c3cf9bb5ccabdb584999d515b53e4757e86610d5/example/v1/tests/stableconfigtypes.example.openshift.io/AAA_ungated.yaml#L258-L281

Addressed by adding ratchets test

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec:  
+          containerRuntimeConfig:
+            logLevel: info
+      expected: |
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec: 
+          containerRuntimeConfig:
```
Comment (everettraven) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: Should `defaultRuntime` be set to the empty string value here?

Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: @everettraven I've reverted the omitempty removal and kept defaultRuntime as a non-pointer type. 

My reasoning is:

- MCO handles both unset and empty string values for defaultRuntime as a default to crun.
- Changing it to a pointer would require extensive code changes. May be across MCO and API repos.
- Keeping `omitempty` ensures the field is omitted when unset, avoiding a confusing empty string for users.
- If a user explicitly sets it to an empty string, it will be included in the manifest, which is acceptable as MCO handles it.

File: machineconfiguration/v1/types.go (hunk 4 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go:855]: Reverted change, See comment : https://github.com/openshift/api/pull/2370#discussion_r2167902778

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 6 above)
Comment (JoelSpeed) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: > MCO handles both unset and empty string values for defaultRuntime as a default to crun.

Without a pointer, MCO cannot tell the difference between unset and the empty string.

Does the empty string mean anything to crun?

> Keeping omitempty ensures the field is omitted when unset, avoiding a confusing empty string for users.

This is called discoverability in our APIs. When a controller generates a configuration API object, we try to generate a complete set of fields so that a user can easily see from the object, where there are options they may not have known about (they can discover the new options).

> If a user explicitly sets it to an empty string, it will be included in the manifest, which is acceptable as MCO handles it.

But it won't round trip. As soon as a structured client writes the object, the field will be removed from etcd and future reads will not render the `""` value. This means that an unstructured apply would never reach a stable state when the controller is responding to the writes from the unstructured client.

Either remove `omitempty`, or remove `""` from the enum of valid values. Removing an enum value is a breaking change btw, so if this is already shipped and the enum already allows `""`, you should remove `omitempty` as this is the least breaking change.

Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: Thanks @JoelSpeed . I have removed `omitempty` and updated test cases.

Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: @JoelSpeed can you review? Happy to address any comments you have

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml [hunk 7]
```diff
@@ -3,7 +3,7 @@ name: "ContainerRuntimeConfig"
 crdName: containerruntimeconfigs.machineconfiguration.openshift.io
 tests:
   onCreate:
-    - name: Should be able to create a minimal ContainerRuntimeConfig
+    - name: Should be able to create a minimal ContainerRuntimeConfig and default runtime set
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml]: Don't modify the existing case, add a new case

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 6 above)
Comment (JoelSpeed) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:25]: These are unstructured calls, unless there's an explicit default, I wouldn't expect the `defaultRuntime` to be added as part of this test

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec:
+          containerRuntimeConfig:
+            defaultRuntime: ""
+            logLevel: info
+    - name: Should be able to update from runc to crun with other parameters
+      initialCRDPatches:
+        - op: remove
+          path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/containerRuntimeConfig/properties/defaultRuntime/enum
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:108]: Why remove the enum on this one? runc is a valid choice no?

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+        kind: ContainerRuntimeConfig
+        spec: 
+          containerRuntimeConfig:
+            defaultRuntime: runc
+            logLevel: info
+  onUpdate:
+    - name: Should not be able to update from default to invalid value
+      initialCRDPatches:
+        - op: remove
+          path: /spec/versions/0/schema/openAPIV3Schema/properties/spec/properties/containerRuntimeConfig/properties/defaultRuntime/enum
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml:83]: Why remove the enum on thius one, "" is a valid choice no?

File: machineconfiguration/v1/types.go [hunk 10]
```diff
@@ -844,18 +844,26 @@ type ContainerRuntimeConfiguration struct {
...
-	// +optional
-	DefaultRuntime ContainerRuntimeDefaultRuntime `json:"defaultRuntime,omitempty"`
+	// defaultRuntime is the name of the OCI runtime to be used as the default for containers.
+	// Allowed values are `runc`, `crun`, and omitted (`""`).
+	// When set to `runc`, OpenShift will use runc to execute the container
+	// When set to `crun`, OpenShift will use crun to execute the container
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default,
+	// which is subject to change over time. Currently, the default is `crun`.
+	// +kubebuilder:validation:Enum=crun;runc;""
+	// +kubebuilder:default:=""
```
Comment (JoelSpeed) [inline machineconfiguration/v1/types.go]: So this is interesting, this creates discoverability even from unstructured clients, we haven't done this before

What's the motivation?

File: machineconfiguration/v1/types.go [hunk 11]
```diff
@@ -844,18 +844,26 @@ type ContainerRuntimeConfiguration struct {
...
-	// defaultRuntime is the name of the OCI runtime to be used as the default.
-	// +optional
-	DefaultRuntime ContainerRuntimeDefaultRuntime `json:"defaultRuntime,omitempty"`
+	// defaultRuntime is the name of the OCI runtime to be used as the default for containers.
+	// Allowed values are `runc`, `crun`, and omitted (`""`).
+	// When set to `runc`, OpenShift will use runc to execute the container
+	// When set to `crun`, OpenShift will use crun to execute the container
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default,
+	// which is subject to change over time. Currently, the default is `crun`.
+	// +kubebuilder:validation:Enum=crun;runc;""
```
Comment (JoelSpeed) [inline machineconfiguration/v1/types.go]: I'm not convinced we actually want to add `""` to the enum here. Are we aiming for discoverability? Who creates this object? Is it generally created by a controller/structured client? Or do we expect end users to create these?

Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go]: @JoelSpeed  Before diving into other comments, let's clarify the `defaultRuntime` field. Our primary goal is to inform users that only 'runc' and 'crun' are valid options, as they currently aren't aware. We also expect users to directly modify this field to change the node's underlying runtime.

Removing omitempty and handling `""` as an invalid value could break existing user configurations if they previously set defaultRuntime to `""`.

Given this, I see these paths forward:

1. Skip Enum Validation for now: Document 'runc' and 'crun' as the allowed values, and let MCO manage it as before. 
2. Remove `""` and omitempty (breaking change): Proceed with stricter validation, accepting this might break a small number of existing users who explicitly set ""."
3. The current change which uses `""` as an enum option so that the existing users aren't broken.

Comment (everettraven) [reply machineconfiguration/v1/types.go]: Joel and I discussed this a bit further offline, and it looks like I missed the nuance of this field already having been set as `omitempty` previously meaning that explicitly setting the empty value (`""`) and completely omitting the field are treated exactly the same by the kube-apiserver (i.e users aren't broken if the enum doesn't include `""`).

That means we should:
- Remove the `""` value from the enum
- Remove the default marker
- Add `omitempty` back
- Keep the messaging about default behavior when omitted in the GoDoc

My apologies for the confusion here.

Comment (JoelSpeed) [reply machineconfiguration/v1/types.go]: @ngopalak-redhat What is the difference between this field being omitted (not present at all), and `""`? Is there a semantic difference?

Comment (JoelSpeed) [reply machineconfiguration/v1/types.go]: As far as I can tell from the docs, it mentions only `runc` and `crun` as valid values, https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/machine_configuration/machine-configs-custom

Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go]: @everettraven When we remove the `""` option in the enum and keep omitempty, it leads to an error if a customer has previously set defaultRuntime to "". You can see the specific error below:
```
                     Type: "FieldValueNotSupported",
                           Message: "Unsupported value: \"\": supported values: \"crun\", \"runc\"",
                           Field: "spec.containerRuntimeConfig.defaultRuntime",
```
My take is that this breakage should be acceptable. We've never actually instructed customers to set defaultRuntime to an empty string. Therefore, any existing configurations with "" are likely not intended
What are your thoughts?

@JoelSpeed  crun is the default now. Omitting or "" for defaultRuntime will result in choosing crun.

Comment (JoelSpeed) [reply machineconfiguration/v1/types.go]: > it leads to an error if a customer has previously set defaultRuntime to ""

Which version of OpenShift did you test this with? Ratcheting configuration should allow the previous value to be persisted on the latest few versions of OpenShift.

> Therefore, any existing configurations with "" are likely not intended
What are your thoughts?

My concern is that, the way the field is structured today, allows an unstructured client (think kubectl apply) to set the field to `""`, but a structured client (think the controller) who updates the object, will strip the value (because of the omitempty). This means that `""` doesn't round trip currently, and anyone who _has_ tried to set the value to `""`, I expect has seen it removed after they have tried to set it. In theory that could be tested.

> @JoelSpeed crun is the default now. Omitting or "" for defaultRuntime will result in choosing crun.

This implies to me that the value `""` should not be in the enum. Though please don't update right now, @deads2k had some comments about this and I'll talk this through with him tomorrow

Comment (JoelSpeed) [reply machineconfiguration/v1/types.go]: Lets continue with this and remove `""`. We can't see anywhere that we told users to do this, and if they did, it wouldn't round trip through a structured client today either

Comment (ngopalak-redhat) [reply machineconfiguration/v1/types.go]: @JoelSpeed I have removed `""`. Updated the PR. Please review

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml [hunk 12]
```diff
@@ -13,4 +13,4 @@ tests:
         apiVersion: machineconfiguration.openshift.io/v1
         kind: ContainerRuntimeConfig
         spec: 
-          containerRuntimeConfig: {}
+          containerRuntimeConfig: {}
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml]: Nit, you've lost the newline char at the end of this file somehow

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 13]
```diff
@@ -0,0 +1,166 @@
...
+      initial: |
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec:
+          containerRuntimeConfig:
+            defaultRuntime: crun
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: The point of removing the enum in the ractheting tests is to set the value to something invalid. Set this to `docker` in initial`, then create three different updated, One to change another field and show it remains, one to show it can be removed, one to update it to a valid value.

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 14]
```diff
@@ -0,0 +1,166 @@
...
+        spec:
+          containerRuntimeConfig:
+            logLevel: info
+      expected: |
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec:
+          containerRuntimeConfig:
+            logLevel: info
+    - name: Should be able to update from runc to crun with other parameters
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: You don't need to remove the enum from the initial CRD validation for this test? You're not actually testing ratcheting unless the `initial` value has something that is not valid

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml [hunk 15]
```diff
@@ -0,0 +1,166 @@
...
+          containerRuntimeConfig:
+            defaultRuntime: docker
+            logLevel: info
+      expected: |
+        apiVersion: machineconfiguration.openshift.io/v1
+        kind: ContainerRuntimeConfig
+        spec:
+          containerRuntimeConfig:
+            defaultRuntime: docker
+            logLevel: info
```
Comment (JoelSpeed) [inline machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: This is a correct use of ratcheting, but you also need to show you can transition from an invalid to valid value, and from an invalid value to absent

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml (hunk 12 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/AAA_ungated.yaml]: Fixed

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 13 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: Fixed

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 14 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: Fixed

File: machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml (hunk 15 above)
Comment (ngopalak-redhat) [reply machineconfiguration/v1/tests/containerruntimeconfigs.machineconfiguration.openshift.io/DefaultRuntimeConfig.yaml]: Fixed
